Jenkins upgrade to 2.289.1

nhojpatrick opened the pull request:
## JENKINS-67789/PR-75 - Jenkins upgrade to 2.289.1

Upgrade Jenkins to latest versions and plugins so can start working on `multibranchPipelineJob` support.

See [JENKINS-67789](https://issues.jenkins-ci.org/browse/JENKINS-67789).

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/gogs-webhook-plugin/blob/master/CONTRIBUTING.md) doc
Yes it's a 404 error pages from GitHub
- [ ] I have referenced the Jira/Github issue related to my changes in one or more commit messages
- [ ] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No findbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

What types of changes does your code introduce? _Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Further comments

If this is a relatively large or complex change, start the discussion by explaining why you chose the solution you did and what alternatives you considered.

nhojpatrick committed:
Upgrade to org.jenkins-ci.plugins:plugin:4.34

nhojpatrick committed:
Upgrade to jenkins v2.289.1

(cherry picked from commit 52603d3b0aa314f167f38bf3b0c538f3db3f01bd)

nhojpatrick committed:
Fix junit:junit version issue

nhojpatrick committed:
Upgrade hamcrest imports to org.hamcrest.core package versions

nhojpatrick committed:
Fix ParameterizedJobMixIn.ParameterizedJob.values() signature issue

nhojpatrick committed:
Fix commons-codec:commons-coded to use bom version

(cherry picked from commit d31712015976569568aa0369806cdba616c2726b)

nhojpatrick committed:
Fix com.infradna.tool:bridge-method-annotation version issue

nhojpatrick committed:
Jenkins Plugin Upgrades to get tests passing

jglick commented on .mvn/wrapper/maven-wrapper.properties:
```diff
@@ -0,0 +1,18 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.8.4/apache-maven-3.8.4-bin.zip
```
As you prefer, but beware that this would only be for the benefit of developers working locally; `buildPlugin` on ci.jenkins.io will ignore this.

nhojpatrick replied:
Maybe this is something should change, what if the project needs or want a newer version. It also means you can use jenkins to verify the build works on a newer maven version. Apache Commons also have the same comments but it does mean you have 1 less management setup on your CICD process.

jglick commented on pom.xml:
```diff
@@ -49,22 +49,52 @@
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>git</artifactId>
-      <version>2.2.5</version>
+      <version>3.6.0</version>
```
https://github.com/jenkinsci/bom

nhojpatrick committed:
Use jenkins-test-harness.version from org.jenkins-ci.plugins:plugin

nhojpatrick committed:
Use io.jenkins.tools.bom:bom-2.289.x:961.vf0c9f6f59827

nhojpatrick committed:
Fix org.jenkins-ci.plugins:git to use bom version

(cherry picked from commit 19d521147ca6c3aec7aef9bd73c308c96e663dd7)

nhojpatrick committed:
Fix org.jenkins-ci.plugins:cloudbees-folder to use bom version

(cherry picked from commit 9ad2d1bff3667127d18b81b15d81619218dbda37)

nhojpatrick committed:
Fix org.jenkins-ci.plugins:credentials to use bom version

(cherry picked from commit a6e149ba09530c91d78149d9bee7cd976ed25dbe)

nhojpatrick committed:
Fix org.jenkins-ci.plugins:junit to use bom version

(cherry picked from commit ddbbb9b3729a7b3a8d68ed42876131b4dd43bf3d)

nhojpatrick committed:
Fix org.jenkins-ci.plugins:structs to use bom version

(cherry picked from commit 2e43e8d5ca7c8daafdd288b88cd087f709fb5247)

nhojpatrick commented:
@jglick have switched to inherited value of jenkins-test-harness.version. have also added added io.jenkins.tools.bom:bom-2.289.x.

I've also removed mvnw.
I've also remove the catch all error handling.

nhojpatrick committed:
Fix org.jenkins-ci:symbol-annotation to use bom version

(cherry picked from commit 71f2988a2bf5663f38a039494075079482d93605)

nhojpatrick committed:
Fix org.jenkins-ci:symbol-annotation to use bom version

nhojpatrick committed:
Fix org.jenkins-ci:symbol-annotation to use bom version

(cherry picked from commit bbd4ed2ac776f9d0552e0fbebf3cf7000c43da43)

nhojpatrick committed:
fix-stuff

nhojpatrick committed:
Fix org.jenkins-ci.plugins:git-client to use bom version

(cherry picked from commit 1862f955e56028870fb69f687f78016cd653981b)

nhojpatrick committed:
Fix org.jenkins-ci.plugins:jsch to use bom version

(cherry picked from commit 90d37defc28b232da84d5671f5d97b5ee20a346b)

nhojpatrick committed:
Fix org.jenkins-ci.plugins:jackson2-api to use bom version

(cherry picked from commit 40251bccd7bdd7fbae5ffc58af10266d8e80d50d)

nhojpatrick committed:
Fix com.fasterxml.jackson.core:*:2.12.4:test

nhojpatrick committed:
Fix org.apache.httpcomponents:fluent-hc:4.5.13

nhojpatrick committed:
Fix org.apache.httpcomponents:httpclient:4.5.13

MarkEWaite commented:
@nhojpatrick if you're interested in exploring some other possible simplifications, refer to the commits at https://github.com/MarkEWaite/gogs-webhook-plugin/commits/jenkins-upgrade .  They pass the automated tests and the plugin loads into a Jenkins 2.289.1 session with `mvn hpi:run`.  The dependencies are further simplified and the contents of the plugin `hpi` file are significantly reduced.

They fail the integration tests in a different way than the integration tests fail for me on the master branch.  I didn't spend time investigating the integration test failures.

sanderv32 commented:
@MarkEWaite How much work will it be to get your changes merged in this plugin? I made this plugin years ago but i don't have the time myself to maintain it and would like to have some help to maintain it and extend it.

MarkEWaite commented:
On Tue, Feb 22, 2022 at 1:30 AM Alexander ***@***.***> wrote:

> @MarkEWaite <https://github.com/MarkEWaite> How much work will it be to
> get your changes merged in this plugin? I made this plugin years ago but i
> don't have the time myself to maintain it and would like to have some help
> to maintain it and extend it.
>

I think that the merge is straightforward, though it would be nice if
someone performed some end to end exploratory testing to confirm that there
were no surprises for the user.  Automated tests are passing and the
changes seemed low impact to me when I made them.

> Message ID: ***@***.***>
>

nhojpatrick commented:
@sanderv32 @MarkEWaite so my branch is now passing, assuming PR does the integration tests. otherwise please tell me what command i need and i'll check locally.
I've review your branch and mine, I can do an annotation changes if needed.
Some of the items I've added, i only did so both `mvn test` and `mvn hpi:run`.
I'll go though each difference and see the bare mine pom i need.

nhojpatrick committed:
Switch to using jenkins plugin apache-httpcomponents-client-4-api

MarkEWaite commented:
> @sanderv32 @MarkEWaite so my branch is now passing, assuming PR does the integration tests. otherwise please tell me what command i need and i'll check locally.

I'm confident that the integration tests are not run from ci.jenkins.io.

When I ran the integration tests (based on instructions in one of the files in the repo), they failed on the master branch (one failure mode) and failed on my copy of your changes (different failure mode).  Since they were failing on the master branch, I didn't spend time investigating further.

nhojpatrick committed:
Use jenkins-war from parent/bom/dependencies

(cherry picked from commit 1b1210f0f39de2ee3acf8020ef99b7259df39d9a)

nhojpatrick committed:
Use jenkins-client v0.3.8

nhojpatrick committed:
Use symbol-annotation from parent/bom/dependencies

nhojpatrick committed:
Use jackson from parent/bom/dependencies

nhojpatrick committed:
Break jenkins plugin 'git' dependency

nhojpatrick committed:
Break jenkins plugin 'git-client' dependency

nhojpatrick commented:
So diff is now only
`import edu.umd.cs.findbugs.annotations.NonNull;`
vs
`import javax.annotation.Nonnull;`

Marks version has custom methods vs `import static com.google.common.base.Preconditions.checkNotNull;`

Any preference in using which non null check and checkNotNull?

nhojpatrick commented:
So integration tests, i'm getting these two errors, are they what you are getting.

1) Unexpected error creating GOGS repo: Repository creation call did not return the expected value (returned 401)

2) Exception caught during execution of push command
-> Source ref refs/heads/master doesn't resolve to any object.

MarkEWaite commented:
> So diff is now only `import edu.umd.cs.findbugs.annotations.NonNull;` vs `import javax.annotation.Nonnull;`
> 
> Mark's version has custom methods vs `import static com.google.common.base.Preconditions.checkNotNull;`
> 
> Any preference in using which non null check and checkNotNull?

Very strong preference for the `edu.umd.cs.findbugs.annotations.NonNull`.  Rationale is explained in the [commit message](https://github.com/MarkEWaite/gogs-webhook-plugin/commit/555a93e54f7467f946f2338b0de74074e915035d).  JSR-305 annotations were never adopted in Java.  They are not maintained.  The spotbugs annotations are maintained in the `edu.umd.cs.findbugs.annotations` package.  The JSR-305 annotations have been removed from core and are steadily being removed from plugins.

Relatively strong preference to remove Guava references.  Rationale is explained in the [commit message](https://github.com/MarkEWaite/gogs-webhook-plugin/commit/6cf7c93b987dc7100c5142d4b0753ca0d413d4e0).  Guava 11 was originally used in Jenkins, but then the Guava API evolved with compatibility expectations that are different than the Jenkins compatibility expectations.  Nothing wrong with either project, they were just different compatibility requirements.  Guava has been reduced and removed from Jenkins core and from plugins so that we could upgrade to a current version of Guava.  Replacing Guava dependencies with something else will reduce the risk of later complexities from Guava.

MarkEWaite commented:
> So integration tests, i'm getting these two errors, are they what you are getting.
> 
> 1. Unexpected error creating GOGS repo: Repository creation call did not return the expected value (returned 401)
> 2. Exception caught during execution of push command
>    -> Source ref refs/heads/master doesn't resolve to any object.

I get the following output (running Java 8u322 on Debian testing with Docker ):

```
Running org.jenkinsci.plugins.gogs.GogsWebHook_IT
Tests run: 2, Failures: 1, Errors: 1, Skipped: 0, Time elapsed: 12.463 sec <<< FAILURE!
smokeTest_build_testBranch(org.jenkinsci.plugins.gogs.GogsWebHook_IT)  Time elapsed: 5.351 sec  <<< FAILURE!
java.lang.AssertionError: Unexpected error creating GOGS repo: Repository creation call did not return the expected value (returned 401)
        at org.junit.Assert.fail(Assert.java:89)
        at org.jenkinsci.plugins.gogs.GogsWebHook_IT.doItTest(GogsWebHook_IT.java:121)
        at org.jenkinsci.plugins.gogs.GogsWebHook_IT.smokeTest_build_testBranch(GogsWebHook_IT.java:84)
...
smokeTest_build_masterBranch(org.jenkinsci.plugins.gogs.GogsWebHook_IT)  Time elapsed: 7.077 sec  <<< ERROR!
org.apache.http.client.HttpResponseException: status code: 403, reason phrase: Forbidden
        at com.offbytwo.jenkins.client.validator.HttpResponseValidator.validateResponse(HttpResponseValidator.java:11)
        at com.offbytwo.jenkins.client.JenkinsHttpClient.post_xml(JenkinsHttpClient.java:342)
        at com.offbytwo.jenkins.JenkinsServer.createJob(JenkinsServer.java:382)
        at com.offbytwo.jenkins.JenkinsServer.createJob(JenkinsServer.java:338)
        at org.jenkinsci.plugins.gogs.GogsWebHook_IT.doItTest(GogsWebHook_IT.java:174)
        at org.jenkinsci.plugins.gogs.GogsWebHook_IT.smokeTest_build_masterBranch(GogsWebHook_IT.java:74)

Results :

Failed tests:
  GogsWebHook_IT.smokeTest_build_testBranch:84->doItTest:121 Unexpected error creating GOGS repo: Repository creation call did not return the expected value (returned 401)

Tests in error:
  GogsWebHook_IT.smokeTest_build_masterBranch:74->doItTest:174 » HttpResponse st...

Tests run: 2, Failures: 1, Errors: 1, Skipped: 0
```

MarkEWaite commented:
> So integration tests, i'm getting these two errors, are they what you are getting.
> 
> 1. Unexpected error creating GOGS repo: Repository creation call did not return the expected value (returned 401)
> 2. Exception caught during execution of push command
>    -> Source ref refs/heads/master doesn't resolve to any object.

I get the following output on the master branch (running Java 8u322 on Debian testing with Docker 20.10.11+dfsg1, build dea9396):

```
Running org.jenkinsci.plugins.gogs.GogsWebHook_IT
Tests run: 2, Failures: 1, Errors: 1, Skipped: 0, Time elapsed: 12.806 sec <<< FAILURE!
smokeTest_build_testBranch(org.jenkinsci.plugins.gogs.GogsWebHook_IT)  Time elapsed: 5.754 sec  <<< FAILURE!
java.lang.AssertionError: Unexpected error creating GOGS repo: Repository creation call did not return the expected value (returned 401)
        at org.junit.Assert.fail(Assert.java:88)
        at org.jenkinsci.plugins.gogs.GogsWebHook_IT.doItTest(GogsWebHook_IT.java:121)
        at org.jenkinsci.plugins.gogs.GogsWebHook_IT.smokeTest_build_testBranch(GogsWebHook_IT.java:84)

smokeTest_build_masterBranch(org.jenkinsci.plugins.gogs.GogsWebHook_IT)  Time elapsed: 6.906 sec  <<< ERROR!
org.apache.http.client.HttpResponseException: Forbidden
        at com.offbytwo.jenkins.client.validator.HttpResponseValidator.validateResponse(HttpResponseValidator.java:11)
        at com.offbytwo.jenkins.client.JenkinsHttpClient.post_xml(JenkinsHttpClient.java:342)
        at com.offbytwo.jenkins.JenkinsServer.createJob(JenkinsServer.java:382)
        at com.offbytwo.jenkins.JenkinsServer.createJob(JenkinsServer.java:338)
        at org.jenkinsci.plugins.gogs.GogsWebHook_IT.doItTest(GogsWebHook_IT.java:174)
        at org.jenkinsci.plugins.gogs.GogsWebHook_IT.smokeTest_build_masterBranch(GogsWebHook_IT.java:74)


Results :

Failed tests:
  GogsWebHook_IT.smokeTest_build_testBranch:84->doItTest:121 Unexpected error creating GOGS repo: Repository creation call did not return the expected value (returned 401)

Tests in error:
  GogsWebHook_IT.smokeTest_build_masterBranch:74->doItTest:174 » HttpResponse Fo...

Tests run: 2, Failures: 1, Errors: 1, Skipped: 0
```

nhojpatrick commented:
@MarkEWaite I think it's csrf protection related.

1) mvn clean pre-integration-test -P withIntegrationTest
2) jenkins /script execute the following
```
import hudson.security.csrf.DefaultCrumbIssuer
import jenkins.model.Jenkins
if (!Jenkins.instance.isQuietingDown()) {
    def j = Jenkins.instance
    if (j.getCrumbIssuer() != null) {
        j.setCrumbIssuer(null)
        j.save()
    }
}
```
3) mvn failsafe:integration-test  -P withIntegrationTest

I appear to get further on my `jenkins-update` branch. Well the stacktrace appears to show move lines in `org.jenkinsci.plugins.gogs.GogsWebHook_IT.doItTest` have been executed.

But I've had nighmare with csrf recently so will need help.

MarkEWaite commented:
> But I've had nighmare with csrf recently so will need help.

Unfortunately, my experiences with automated end to end tests has lead me to use a completely different technique.  I maintain a [Jenkins instance with configuration as code](https://github.com/MarkEWaite/docker-lfs/tree/lts-with-plugins).  I define jobs in that Jenkins instance that I can use to interactively check status and perform experiments while relying on my other production servers (like my Gitea server). 

I've not had much success with automated end to end tests.  I'm truly impressed by the techniques used in this repository, but I'm not ready to spend time debugging those types of tests.

jglick commented:
https://github.com/jenkinsci/gogs-webhook-plugin/blob/master/about_integration_tests.md#integration-test-internals sounds rather complicated. Consider using `RealJenkinsRule` plus Testcontainers for something that is more IDE-friendly (not reliant on Maven). Automated end-to-end tests are a real pain when dealing with SaaS or licensed products (you have to pick between WireMock or requiring authentication to run tests), but since Gogs is OSS it should be fine.

If tests fail when CSRF protection is enabled, the test might be wrong, but the code being tested might also be wrong. Do not disable CSRF protection as a workaround.

nhojpatrick commented:
@jglick this is my 1st jenkins integration tests, and until @MarkEWaite highlighted they where failed I didn't know they event existed. I've just been doing `rm -rf work/plugins && mvn clean install && mvn hpi:run`.
I'm only trying to upgrade so I can submit a pr so that this plugin supports gogs calling back to a pipeline multibranch job.

Do you know another plugin, maybe a bitbucket or github, where it stands up a git repo server. As the tests does a commit and pushes it to gogs, which then fires the webhook, and the the tests detects the job got triggered. That is my current working assumption regarding these IT's.

jglick commented:
> Do you know another plugin, maybe a bitbucket or github, where it stands up a git repo server.

Neither of those could, at least not easily, due to the proprietary nature (though it seems you can make a BB Server test instance with a trial license or something). They both use WireMock.

I do not know anything about this `com.offbytwo.jenkins.client.JenkinsHttpClient`. Normally plugins make HTTP requests to the Jenkins controller in functional tests using `JenkinsRule.WebClient`.

nhojpatrick commented:
So I've been able to fix 1 of the `GogsWebHook_IT` tests by simply changing;
`jobAtIntitalState.build();`

to;
`jobAtIntitalState.build(true);`

nhojpatrick committed:
Jenkins crumb fix for GogsWebHook_IT

nhojpatrick commented:
So i'm moving forward and backwards, now i need to move `~/.ssh` and `~/.gitconfig_global` otherwise i get other errors. One example is i've a global `init.defaultbranch=main` but test assumes it's called master.

MarkEWaite commented:
> So i'm moving forward and backwards, now i need to move `~/.ssh` and `~/.gitconfig_global` otherwise i get other errors. One example is i've a global `init.defaultbranch=main` but test assumes it's called master.

That's the same complication I've had to manage in the git client plugin automated tests and the git plugin automated tests.  State that is outside the repository (in this case, the user's git configuration) affects the behavior of the test.  On positive side, that means the test is exploring things that are nearer to real users.  On the negative side, it means that the tests need to be extended to handle multiple configurations.

jglick commented:
> State that is outside the repository (in this case, the user's git configuration) affects the behavior of the test.

Something best handled with Testcontainers. (At least for code run on agents. Currently `RealJenkinsRule` does not support running the controller in a defined container, but I think this could be added.)

MarkEWaite commented on pom.xml:
```diff
@@ -79,30 +117,30 @@
       <groupId>org.eclipse.jgit</groupId>
       <artifactId>org.eclipse.jgit</artifactId>
-      <version>4.7.0.201704051617-r</version>
+      <version>5.12.0.202106070339-r</version>
```
The most recent version supported by the git client plugin is:

Since this is a test dependency, you can leave it as the older version if you wish.

Can't make it depend on JGit 6.0 because that does not compile with Java 8.
```suggestion
      <version>5.13.0.202109080827-r</version>
```

MarkEWaite reviewed (approved):
Looks good to me.

nhojpatrick committed:
org.eclipse.jgit v5.13.0.202109080827-r

nhojpatrick commented:
@jglick @MarkEWaite i think it's all good to be merged. As the change I was actually wanted to make was in the logic so it handles multibranchPipelineJob and triggers sub jobs when that branch is changed.

nhojpatrick committed:
Maven xml version, xml encoding and https for xsd

MarkEWaite reviewed (approved):
No objections from me on this pull request.

sanderv32 commented:
LGTM also
